semantic-core/src/Data/Term.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DeriveTraversable, FlexibleInstances, LambdaCase, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Data.Term
+( Term(..)
+, Syntax(..)
+, iter
+, cata
+, interpret
+) where
+
+import Control.Effect.Carrier
+import Control.Monad (ap)
+import Control.Monad.Module
+import Data.Coerce (coerce)
+import Data.Functor.Const
+import Data.Scope
+
+data Term sig a
+  = Var a
+  | Term (sig (Term sig) a)
+
+deriving instance ( Eq a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  )
+               => Eq  (Term sig a)
+deriving instance ( Ord a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  , forall g x . (Ord x, Monad g, forall y . Eq  y => Eq  (g y)
+                                                , forall y . Ord y => Ord (g y)) => Ord (sig g x)
```

Comment:
…now with horrifying higher-order quantified constraints!

1. I didn’t want to use `Eq1`, `Ord1`, `Show1`, etc., because they’re awful, and don’t provide derivations.

2. `Core` (and any other syntax which binds variables) applies its type parameter `f` to more than one different type: `f a` for normal subterm positions` and `Scope () f a` (isomorphic to `f (Incr () (f a))`) for binders like `Lam`. Therefore, `Core` uses quantified constraints for its `Eq` instance to state that it can be equated whenever `f x` can be equated for all `x` that can be equated.

3. `Term` only ever applies its type parameter `sig` to `Term sig` and `a`; however, presumably due to `Core`’s quantified constraints, using the simpler constraint of e.g. `(Show a, forall x . Show x => Show (sig (Term sig) x))` causes the constraint solver to infinite loop. `UndecidableInstances` means business in this case. Therefore, we have to use the dramatically more complicated constraint `(Show a, forall g x . (Show x, forall y . Show y => Show (g y)) => Show (sig g x))`, which quantifies over some arbitrary type parameter `g`.

4. Both the `Eq` & `Ord` instances require `Monad` instances for `g` because `Core` requires a `Monad` instance for its parameter to be equated (due to the use of `Scope`). Likewise, they both require a `RightModule` instance for `sig`, because `Term` requires a `RightModule` instance for `sig` in its `Monad` instance. So these instances compile fine if you :fire: the `RightModule` & `Monad` constraints, but then they can’t actually be instantiated to work with `Core`.

5. The `Ord` instance duplicates the quantified `Eq` constraint (just like `Core`’s does) because of contravariance: `forall x . Ord x => c` is strictly smaller than `forall x . Eq x => c`, because there are fewer `Ord` instances than `Eq` instances. Therefore, it’s not possible to prove `forall x . Eq x => Eq (f x)` from `forall x . Ord x => Ord (f x)`, because the domain of the smaller is too small to prove its premise.